### PR TITLE
Container & instead of smart pointers

### LINKED ISSUE #226: Skippable tuning phases
**Is your feature request related to a problem? Please describe.**
Currently, either tuning happens very often or maybe very seldom.
It would be nice if AutoPas were able to decide itself when to tune.

**Describe the solution you'd like**
The easiest way to decide this would probably to allow skippable tuning in a specific interval, e.g. every 1000 time steps. Such that AutoPas could decide whether it actually wants to tune or whether it doesn't.

**Additional context**
We have to take care that the tuning doesn't destroy a potential sync between the rebuilding of the containers/lists.

### LINKED ISSUE #720: Tuning strategy overhaul
**Is your feature request related to a problem? Please describe.**
The current framework for tuning is rather inflexible and makes combining ideas from different tuning strategies difficult. We want to come up with a new framework which provides the flexibility to develop tuning further.

**Describe the solution you'd like**
TBD

### PR DESCRIPTION
**Pull request by FG-TUM**
# Description

- [x] Pass container references instead of smart pointers. (Implements [Core Guideline F.7](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f7-for-general-use-take-t-or-t-arguments-rather-than-smart-pointers))
- [x] some more `const`
- [x] minor variable and function renaming

## Related Pull Requests

- Already contains #678

### CODE REVIEW

**Review by thesamriel — COMMENTED** [review]
PR looks good. Only minor points.
Maybe you could write down the reasons or ideas behind this change.

**Comment by FG-TUM** [comment]
> Maybe you could write down the reasons or ideas behind this change.

I wanted simpler interfaces, and code. This also implements the [Cpp Code Guidleine F.7](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f7-for-general-use-take-t-or-t-arguments-rather-than-smart-pointers).

File: src/autopas/selectors/AutoTuner.h [hunk 1]
```diff
@@ -868,15 +857,17 @@ void AutoTuner<Particle>::iteratePairwiseTemplateHelper(PairwiseFunctor *f, bool
...
   timerIteratePairwise.start();
-  containerPtr->iteratePairwise(traversal.get());
+  container.iteratePairwise(traversal.get());
   timerIteratePairwise.stop();
 
   timerRemainderTraversal.start();
-  doRemainderTraversal<useNewton3>(f, containerPtr, particleBuffer, haloParticleBuffer);
+  withStaticContainerType(container, [&](auto &actualContainerType) {
+    doRemainderTraversal<useNewton3>(f, actualContainerType, particleBuffer, haloParticleBuffer);
+  });
```
[the remark this answers is not in the document]
Comment (FG-TUM) [reply src/autopas/selectors/AutoTuner.h:870]: The problem is, that you can't have virtual functions that are templated. See [here](https://stackoverflow.com/a/22422889/7019073) for more context.

File: tests/testAutopas/tests/autopasInterface/Newton3OnOffTest.cpp [hunk 2]
```diff
@@ -10,6 +10,7 @@
 #include "autopas/selectors/TraversalSelector.h"
 #include "autopas/utils/StaticCellSelector.h"
 #include "autopas/utils/logging/Logger.h"
+#include "testingHelpers/commonTypedefs.h"
```
[the remark this answers is not in the document]
Comment (FG-TUM) [reply tests/testAutopas/tests/autopasInterface/Newton3OnOffTest.cpp]: fixed 5b505d4

File: src/autopas/utils/StaticContainerSelector.h [hunk 3]
```diff
@@ -30,34 +30,33 @@ namespace autopas {
  * @return Returns whatever function returns.
  */
 template <typename Particle, typename FunctionType>
-decltype(auto) withStaticContainerType(const std::shared_ptr<ParticleContainerInterface<Particle>> &container,
+decltype(auto) withStaticContainerType(autopas::ParticleContainerInterface<Particle> &container,
```
[the remark this answers is not in the document]
Comment (FG-TUM) [reply src/autopas/utils/StaticContainerSelector.h:33]: fixed 9f9c0cd

File: src/autopas/selectors/AutoTuner.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (thesamriel) [reply src/autopas/selectors/AutoTuner.h:870]: I see, that is unfortunately a valid argument

**Review by thesamriel — APPROVED** [review]

**Pull request merged**